ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java

```diff
@@ -2614,28 +2633,25 @@ public static Path getVersionFilePath(Path deltaOrBase) {
           + " from " + jc.get(ValidTxnWriteIdList.VALID_TABLES_WRITEIDS_KEY));
       return null;
     }
-    Directory acidInfo = AcidUtils.getAcidState(fs, dir, jc, idList, null, false);
+    if (fs == null) {
+      fs = dir.getFileSystem(jc);
+    }
+    // Collect the all of the files/dirs
+    Map<Path, HdfsDirSnapshot> hdfsDirSnapshots = AcidUtils.getHdfsDirSnapshots(fs, dir);
```

Comment:
this might be out-of-scope for this change: but this *static* method in `AcidUtils` is trying to do all the work upfront...
which might lead to:
* that it does work which is not even needed
* it doesn't scan some location - and the map just returns null ; so it might be not noticable

I think it would be better if this method would return a something (it could still be a map) which could fill in stuff from hdfs if its not cached already...

Reply:
Ohh.. I think I get it now.
* You are right that this will do stuff which is not really needed in this case - namely creating objects which are not needed at here (dirSnapshot.metaDataFile/dirSnapshot.acidFormatFile), also we might list and create objects which are not needed in this snapshot. On the other hand the costly part on S3 (and on HDFS as well) is the number of remote calls, which is reduced to a single listing instead of doing the listing for every directory 1-by-1.
* It is not possible that it does not scan some location which needed. If this happens then this is a bug in AcidUtils.getAcidState, as it has to return every directory which is readable

What I do not understand in your comment is "this method would return a something (it could still be a map) which could fill in stuff from hdfs if its not cached already" - the main thing we would like to avoid here is the need of reading the HDFS again and again. The only way to realize that something is missing is reading the directory again... or I miss something :)